schism/page_patedit.c

```diff
@@ -2958,28 +2958,41 @@ static int patedit_record_note(song_note_t *cur_note, int channel, UNUSED int ro
 static int pattern_editor_insert_midi(struct key_event *k)
 {
 	song_note_t *pattern, *cur_note = NULL;
-	int n, v = 0, c = 0, pd, speed, tick;
+	int n, v = 0, pd, speed, tick;
+	int r = current_row, c = current_channel;
+	int first_note = 0;
+	int quantize_next_row = 0;
 
 	status.flags |= SONG_NEEDS_SAVE;
 	song_get_pattern(current_pattern, &pattern);
 
 	if (midi_start_record && !(song_get_mode() & (MODE_PLAYING|MODE_PATTERN_LOOP))) {
 		switch (midi_start_record) {
 		case 1: /* pattern loop */
-			song_loop_pattern(current_pattern, current_row);
+			song_loop_pattern(current_pattern, r);
 			midi_playback_tracing = playback_tracing;
 			playback_tracing = 1;
 			break;
 		case 2: /* song play */
-			song_start_at_pattern(current_pattern, current_row);
+			song_start_at_pattern(current_pattern, r);
 			midi_playback_tracing = playback_tracing;
 			playback_tracing = 1;
 			break;
 		};
+
+		first_note = 1;
 	}
 
 	speed = song_get_current_speed();
 	tick = song_get_current_tick();
+
+	/* correct late notes to the next row */
+	/* tick + 1 because processing the keydown itself takes another tick */
+	if (midi_flags & MIDI_TICK_QUANTIZE && !first_note && tick + 1 < speed / 2) {
+		r = (r + 1) % song_get_rows_in_pattern(current_pattern);
+		quantize_next_row = 1;
+	}
+
```

Comment:
If a MIDI note is entered on the last row of a pattern and should be quantized to the next row, does it just write to the first row of the current pattern then?

Reply:
well.. yeah. What other behavior do you prefer?
* clamp row to max-row
* write note to next pattern

Reply:
Ideally it would write to the next row that would be played, which requires some nontrivial logic (which could hopefully be factored out of wherever it exists in the codebase already):

- If the current pattern is being looped, the correct behavior is to write to the first row of the current pattern
- If the song is being played normally, the correct behavior is to write to the first row of the next pattern in the order list
- Jump/break commands (Bxx, Cxx, SBx) should be correctly accounted for (in general, not just at the end of a pattern)

Reply:
very interesting! will work on this soon.